### ISSUE TITLE
ADCIRC integration

### ISSUE THREAD

**Issue by uturuncoglu** [oceanmodeling/ufs-weather-model#4]
I am creating new issue for ADCIRC integration. As we discussed today, I'll look at option to create additional layer in the ADCIRC cap to interact with CMEPS. I'll update you about it once I have discussion with ESMF core team.

**Comment by uturuncoglu** [comment]
@pvelissariou1 I wonder if there is any progress about the issue that we faced during writing output? I think you were plaining to fix the configuration files and maybe you have some update about it.

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***>
I modified the fort.15 files in the run folders for the ADCIRC cases to
print out the results every hour
(checked this on hera).
fort.63.nc : Elevation Time Series at All Nodes in the Model Grid
fort.64.nc : Depth-averaged Velocity Time Series at All Nodes in the Model
Grid
fort.73.nc : Atmospheric Pressure Time Series at All Nodes in the Model Grid
fort.74.nc : Wind Stress or Velocity Time Series at All Nodes in the Model
Grid

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 Thanks for your help. I will get new file form test suite repo. Do I need to do anything else? You know, i am using CFS data for this case.

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> Just get the fort.15 file from the run
directory. You don't have to modify fort.15,
it is designed to output data every hour for 7 days.
If you just want to run the model for just one day, just modify the
model_configure file accordingly.
Also, I guess you have modified the config.rc file to point to the CFS data.

Takis

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 In my case model fails in ADCIRC_Init. I'll run the prep from scratch and test again to see what happens.

**Comment by uturuncoglu** [comment]
@pvelissariou1 prep is also giving following error,

`0:forrtl: severe (59): list-directed I/O syntax error, unit 15, file /glade/scratch/turuncu/COASTAL_APP/florence_hsofs.atm2adc/fort.15`

Do I need to update the model? I think this is working for you. Right?

**Comment by uturuncoglu** [comment]
@pvelissariou1 I think I am using wrong file. Let me copy again.

**Comment by uturuncoglu** [comment]
@pvelissariou1 It seems that it is writing to fort.* files but I have still zero time dimension in following files,

```
-rw-r--r-- 1 turuncu ncar  45505303 Mar 24 12:56 maxele.63.nc
-rw-r--r-- 1 turuncu ncar  45505232 Mar 24 12:57 maxrs.63.nc
-rw-r--r-- 1 turuncu ncar  45510050 Mar 24 12:57 maxvel.63.nc
-rw-r--r-- 1 turuncu ncar  45505182 Mar 24 12:57 maxwvel.63.nc
-rw-r--r-- 1 turuncu ncar  45505295 Mar 24 12:57 minpr.63.nc
```

Is it same also for you? Anyway, I think I need to check the model results from fort* files. If you don't mind could you remind me about the content of those files?

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> Sorry for the late reply, I was working
on something else.
You need to copy the fort.15 files in the run folders of the hsofs test
cases. They are working.
You don't need to update/recompile anything.
The files maxele.63.nc  maxvel.63.nc  maxwvel.63.nc  minpr.63.nc should
only have 1 time record
in them (the last time record) as they contain the max/or min values of the
results for the simulation period.
The file maxrs.63.nc is for the wave radiation stresses (no waves in this
simulation).

Please use fort.63.nc and fort.64.nc when comparing the results.



Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 I have no data in maxele.63.nc, maxvel.63.nc, maxwvel.63.nc and minpr.63.nc. Their time dimension is 0. I think this is because we set the simulation length 7 days and I am running model 1 day. Maybe I could run 7 days to see what happens. What do I need to change to adjust fort.15 to run mode 1 day? fort.63 and 64 seems fine at this point.

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> In the fort.15 file there is the RNDAY
line for the total simulation time.
*RNDAY* value should be *7 + run_simulation_days*. The *7* represents the
spinup period used for the simulation.
I have set RNDAY to 14 in the fort.15 files. For shorter simulation periods
you may only adjust the times in the model_configure file.

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 Okay. I understand. I'll try to adjust it for my case. Thanks for your great help. Sorry about lots of message.

**Comment by uturuncoglu** [comment]
I am not sure this is expected or not but I plot fort.63.nc file using `adcircpy` tool. Here is the plot for first time step and last one.

![fort63_0](https://user-images.githubusercontent.com/1228999/230491112-5d3f45be-3647-45e9-8ede-82edabca3828.png)
![fort63_23](https://user-images.githubusercontent.com/1228999/230491123-04a5a151-eafd-49f8-acbb-2e8ea592b29a.png)

The last time step seems suspicious to me. In this case I am running model 1 day (let me know, if I need to run longer). Since we are forcing the model with CFS data, there might be some issue over there. I'll try to create visualization for the fields that goes to ADCIRC to be sure that interpolation is okay.

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> Please let it run for a week
Can you plot the data from frot.73/fort.74?

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 Okay. I run it 7 days and now I have data on those files. I also changed WTIMINC parameter and change the temporal resolution of the data from 3600 to 21600 (only first number) to fit to my CFS data. Let me know if you think that that is wrong. I also checked the results after 7 days and it looks similar to after 1 days. There might be an issue in the CFS forcing. Does ADCIRC also write forcing to the file? Is it possible to change the configuration to write those?

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> Yes it does for the interpolated wind
data to ADCIRC's mesh. These are the files fort.73.nc and fort.74.nc.
Explanation:
fort.73.nc : Atmospheric Pressure Time Series at All Nodes in the Model Grid
fort.74.nc : Wind Stress or Velocity Time Series at All Nodes in the Model
Grid

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 I think I found the issue related with the results. The CFS data that you provide me does not cover the small part in south edge of the domain. Please, look at following plot. The white is the CFS mesh and the red one is ADCIRC. As you can see small part of ADCIRC mesh is in the outside of the CFS mesh. So, if you don't mind could you regenerate the CFS data again but in this time please extend it in little bit more south. I think that will solve the issue that we are seeing at this point.

<img width="777" alt="Screen Shot 2023-04-09 at 12 00 42 AM" src="https://user-images.githubusercontent.com/1228999/230757221-be957edf-cb59-421c-bd2e-8137948b3cb7.png">

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***>    Ok, I will recreate the data. Thank
you for checking.

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> Please get the new Florence CFS data
file on orion at:
/work/noaa/nosofs/pvelissa/for_Ufuk_atm_data/Florence_CFS.nc

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 Okay. Thanks. I'll try with this file. BTW, I noticed that this file is smaller then the previous one. The previous dimension was 264x201 but this one has 161x109. Are you sure that the file covers the bigger area than the previous one.

**Comment by uturuncoglu** [comment]
@pvelissariou1 I think there is a resolution difference but I am not sure why. The new data resolution is 0.5 deg. but previous was 0.2 deg. Do you know why?

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***>  I downloaded the 0.5 degree data. Let
me check the availability of the 0.2 degree data.

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> Please get the
data: /work/noaa/nosofs/pvelissa/for_Ufuk_atm_data/Florence_CFS-0.2x0.2.nc
The spatial resolution is 0.2 degrees and the* temporal resolution is 1 hr*
(don't forget to modify fort.15 to reflect this).

Does the spatial resolution affect how CMEPS does the interpolation?

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 Could you let me know the contact person/s for the ADCIRC cap development. If you know their GitHub name that would be great and we could interact through the issue page. Thanks.

**Comment by uturuncoglu** [comment]
@pvelissariou1 BTW, 0.2 deg. data solve the issue. Thanks for your help.

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> I sent an email to ADCIRC folks to
supply us with the info you asked for. Waiting for their response.

For the 0.2 deg. data I made sure that the atm. grid overlaps the ADCIRC
mesh at least by 2.0 degrees at each side.
For your information, the atm. pressure in the 0.2 deg. data is not the
mean sea level converted pressure (the models expect) but rather
the surface pressure (sea level or ground surface) which, in our case is ok
as we go inland up to 10-m above mean sea level.

Takis



Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> The atmospheric data file is in orion
at: /work/noaa/nosofs/pvelissa/for_Ufuk_atm_data/Wind_Pressure_HWRF_Florence_ExtendedSmooth.nc
This is a high resolution product, file size is ~23GB. Please focus your
attention over ocean (not land/rivers), HWRF does not produce good results
over land.

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 This data does not cover entire simulation for `florence_hsofs.atm2adc` configuration. The configuration starts on 2018-09-07 but data starts from 2018-09-09 06:00:00. So, simulation fails. Any idea how this works in ATMMESH. Maybe it is marked with another dataset.

**Comment by uturuncoglu** [comment]
@dwirasae @pvelissariou1 @saeed-moghimi-noaa As I see from the code, the ADCRIC cap uses following subroutine - `extract_parallel_data_from_mesh_orig()` to collect required information for mesh creation,

https://github.com/uturuncoglu/adcirc/blob/7cf14e22b08a978c7530828cc74436a573a6f3bf/thirdparty/nuopc/adc_mod.F90#L256

This seems to read files in PE directories. Anyway, I'll follow the same approach at this point and once we have working cap with CMEPS. We could modify it to use new method.

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> I uploaded to orion
in /work/noaa/nosofs/pvelissa/for_Ufuk_atm_data the file
files_for_HWRF_Florence_ExtendedSmooth.tar.gz
The archive contains the proper fort.67.nc file for the HWRF atm. data you
are using.
I included the updated fort.15 and model_configure files as well.
Just replace the old files with the new ones and re-run the case.
Hopefully, this will work.
Please, let me know how it goes.

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 Are we changing the dates in this case, rather then changing input files? Are you also plaining to update CoastalApp test suite to reflect this. I am asking because the test under ufs-coastal will be different from the test that is available under test suite.

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> Please change the files in your copy of
the testsuite so both have the same forcing and dates. Don't forget to edit
the config.rc file with the new atmospheric input file.

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
I have no `config.rc` since I am using cdeps for forcing not atmmesh. Anyway, let we run the case with those updated files. Thanks for your help.

**Comment by uturuncoglu** [comment]
@dwirasae It seems that fort.14 and fort.18 are not consistent in terms of used element ids. The node coordinates and element connection table is coming from fort.14 but rest of the information is from fort.18. When I try to find the element id in element connection to find the associated nodes to calculate element coordinates I could not since the element ids are not matching between fort.14 and fort.18 for the same PE. Maybe those files are consistent internally but not each other. If this is correct maybe we need to go with new method which can be seen in `extract_parallel_data_from_mesh()` call. Maybe I am missing something in here. Maybe one holds global and other have local ids and we need to do some conversation. Anyway, let me know what you think?

**Comment by dwirasae** [comment]
Element id in a local fort.14 files (i.e. PEXXXX/fort.14) are actually a local element id (but the node ids are global ).  Ids listed in fort.18 (lines after NELG) correspond to a global element id.  exact_parallel_data_from_mesh() should work fine; however, it is expensive since,  if not mistaken, it  search through a global partmesh.txt file.

**Comment by uturuncoglu** [comment]
@dwirasae Anyway, it seems that there is no way to map local ids to global ones. Right? If so, the only option is to go with `exact_parallel_data_from_mesh()`. I think it is not so important to be slow since it will be run only once in the initialization.

**Comment by uturuncoglu** [comment]
@dwirasae I am not sure but still there are some inconsistencies in this subroutine. Some variables such as node coordinates and element connections are coming from ADCIRC internal data structures and others from files. Again, some of them local and others are global. So, I think we need to find a place that we could find all those information in a consistent manner. Is there any documentation that have detailed information about structure of fort.14 (I think I could find it in the wiki), fort.18 and also global parmesh.txt files? Also having some information about internal ADCIRC variables that are related with the mesh would be great.

**Comment by uturuncoglu** [comment]
@dwirasae I think also node ids returned from fort.14 are local and used to construct ElConnect. In the `read14femesh()` call n1, n2, and n3 (actual read values) are converted to local using the call like `find(node_dict_tmp, n1)`. So, they are not the actual values read from the file. Maybe I am misinterpreting in here. Anyway, I think I started to understand the issue. Let me work on it and see what happens.

**Comment by dwirasae** [comment]
@uturuncoglu.  A link below contains information of (a global) fort.14 : 

https://adcirc.org/home/documentation/users-manual-v52/input-file-descriptions/adcirc-grid-and-boundary-information-file-fort-14

To my knowledge, there is no document for fort.18 and partmesh.txt.  

- In a nutshell,  a number in the i-th line of partmesh.txt corresponds to a PE to which the i-th node belong.  

- fort.18 basically contain a local-to-global map and massage tables. Information below might be useful.

   
    NELG    NELG    NELL  NELL   # NELG -  total no elements in the mesh  # NELL - total no eleement in the local subdomain
    DO I = 1, NELL
         GELID(I)  #  Global element ID of the i-th local element. A negative value indicate a ghost element. 
   END DO
    NBODG    NNG  NNLT NLLR  # NNG - total no nodes in the mesh #NNLT - total no of node in the local subdomain      (residet + ghost nodes)
    DO I = 1, NNLT
       GNID(I) # Global node ID of the i-th local nodes
    END DO
    
    NFLUX   # Flux boundary 
    NETA      # OPEN boundatry
    NSTAE    #  No.  elevation station (to which the time series of elevation are recored)
    NSTAV    #  No.  current statiom
    NSTAM   #  No.  meteorological station
    RES NODE #  Massage table 
     ...
     ... 
    ....
    ....

**Comment by dwirasae** [comment]
@uturuncoglu. Yes, the node ids return from read14femesh() are local. read14femsh() use a function find( , )  to decode a hash key (this feature was introduced in v54).  To find the coordinates of the elements, one could do something like 


    do i1 = 1, the_data%NumEl, 1
         
                 xc(1,i1)  = xc(1,i1) +  sum( vk(1, etov(:, i1) ) )/3.0  ;
                 xc(2,i1) =  xc(2.i1) +  sum( vk(2, etov(:, i1) ))/3.0   ;

    end do

**Comment by uturuncoglu** [comment]
@dwirasae Thanks for your suggestion and help. The element coordinates are fine now but I have issue with filtering nodes that is associated with the non-ghost elements. Is there any easy way to access non-ghost elements and their associated nodes. At this point I am developing a new subroutine that takes the `the_data` user defined data structure and modify it by eliminating ghost elements and keeping only associated nodes but maybe there is an easy way that I don't know. Anyway, please let me know if you have any additional information about it.

**Comment by dwirasae** [comment]
@uturuncoglu. The way you did is probably the best way to extract the nodes associated with non-ghost elements. If you just want the non-ghost nodes, you can get them from the_data%owned_to_present_nodes.

**Comment by uturuncoglu** [comment]
@dwirasae I think having non-ghost nodes are fine since edge elements could have ghost nodes. I think I have issue with mapping global node ids found in the element connection array to the local node ids. Is there any way to access that mapping? Can I assume the the global node ids in the element connections are in a same order in the local node ids?

**Comment by saeed-moghimi-noaa** [comment]
Hi @uturuncoglu 

I am using fort 80 to plot domain decomposition using this code. I am not sure if this could be helpful to you


```

def ReadFort80(dir):
    """
    Read fort.80 file for domain decomposition information
    
    return a dictionary including:
    IMAP_EL_LG
    IMAP_NOD_LG
    IMAP_NOD_GL (negative values for not owned elements)
   
    """
    fdata = open( dir + '/fort.80' ,  'r')
    while True:
    #for  line in fdata.readlines():
        line = fdata.readline()
        if 'Number of processors'     in line: nproc = int(line.split()[0]) 
        if 'Total # elements & nodes' in line: 
            nelem = int(line.split()[0])
            nnode = int(line.split()[1])          
        if 'NWLON, NWLAT'             in line:
            print (line)
            break

    #allocate
    IMAP_NOD_LG  = []
    IMAP_NOD_GL   = np.zeros((nnode,3),dtype=np.int)
    IMAP_EL_LG   = []
    #IMAP_STAE_LG = []
    #IMAP_STAV_LG = []
    #IMAP_STAC_LG = []
    #IMAP_STAM_LG = []
    pe_all        = []
    print ('[info:] read nodes local2global')
    for inp in  range( nproc ):
        line1       = fdata.readline()
        #print line1
        pe          = int(line1.split()[0])
        nnodp       = int(line1.split()[1])                
        nod_res_tot = int(line1.split()[2])                
        pe_all.append(pe)
        tmpa = np.array([])
        proc_read = True
        while proc_read:
           line1 = fdata.readline()
           tmp = np.array([int(v) for v in line1.split()])
           tmpa = np.r_[tmpa,tmp]
           if len(tmpa) == nnodp:
               IMAP_NOD_LG.append(tmpa)
               proc_read = False
      
    print ('[info:] read nodes local2global')
    line1       = fdata.readline()
    #print line1
    for il in range(nnode):
        line1       = fdata.readline()
        node_globa = int(line1.split()[0])
        pe         = int(line1.split()[1])                
        node_local = int(line1.split()[2]) 
        IMAP_NOD_GL[il,:] = node_globa , pe ,  node_local

    print ('[info:] read elements local2global')
    for inp in  range( nproc ):
        line1       = fdata.readline()
        #print line1
        pe          = int(line1.split()[0])   # pe number
        nelmp       = int(line1.split()[1])   # element on pe             

        pe_all.append(pe)
        tmpa = np.array([])
        proc_read = True
        while proc_read:
           line1 = fdata.readline()
           tmp = np.array([int(v) for v in line1.split()])
           tmpa = np.r_[tmpa,tmp]
           if len(tmpa) == nelmp:
               IMAP_EL_LG.append(tmpa)
               proc_read = False
           
    fdata.close() 
    return (dict (nproc = nproc, nelem = nelem, nnode = nnode ,
                  IMAP_EL_LG  = np.array(IMAP_EL_LG) ,
                  IMAP_NOD_LG = np.array(IMAP_NOD_LG),
                  IMAP_NOD_GL = np.array(IMAP_NOD_GL) )) 








def plot_domain_decomp(run_dir):
    import pynmd.models.adcirc.post as adcp 
    import pynmd.plotting.plot_settings as ps
    import matplotlib.pyplot as plt
    decomps = adcp.ReadFort80(run_dir+'/')
    nproc = decomps['nproc']
    IMAP_NOD_LG = decomps['IMAP_NOD_LG'] 
    IMAP_EL_LG  = decomps['IMAP_EL_LG'] 
    IMAP_NOD_GL = decomps['IMAP_NOD_GL']
    x,y,tri = adcp.ReadTri  (run_dir+'/')
     
    fig = plt.figure(figsize=(5,5))
    ax  =  fig.add_subplot(111)
    xp  = []
    yp  = []
    cc = 100 * ps.colors
    mm = 100 * ps.marker
    for ip in range(nproc):
        ind = np.where(IMAP_NOD_GL[:,1] == ip)
        ind = ind[::10]
        ax.scatter(tri.x[ind],tri.y[ind], c = cc[ip],
            s = 0.2 , edgecolors = 'None', marker = mm[ip])
    fig.savefig(run_dir + '/0decomp.png',dpi=450)
    plt.close()
    #print ' > Finished plotting domain decomposition'
    txt1 = ' > Finished plotting domain decomposition ..'
    logf(txt1,log_file)          




```

**Comment by saeed-moghimi-noaa** [comment]
See here some of my very old tools:

https://github.com/moghimis/pynmd/blob/master/models/adcirc/post/adcirc_post.py

**Comment by dwirasae** [comment]
@uturuncoglu. Sorry for a late response. The node ids in the_data%Elconnect(3*(i-1):3*i) are the local node ids of the ith local element.

**Comment by uturuncoglu** [comment]
@dwirasae @saeed-moghimi-noaa Thanks both of you. I am trying to build custom ESMF version to see which node or element causing the issue in the ESMF side. So, this might help me to find the issue in my code. I'll update you about it.

**Comment by uturuncoglu** [comment]
@dwirasae Do you have any information about having negative element ids in the fort.18 files? I was thinking the negative element ids indicate the ghost elements but I think it is something different. Could you clarify it for me? I create element mask by looking its nodes are local or not and that mask also marked some elements with negative id as local element. Any idea?

**Comment by uturuncoglu** [comment]
@dwirasae As I see there are some elements with negative id (means ghost) but all nodes are resident. I am expecting to have positive element id for those elements. Do you have any explanation about it?

**Comment by uturuncoglu** [comment]
@pvelissariou1 i was thinking that I solved the issue related to the ADCIRC mesh to make it compatible with CMEPS but apparently not. I have still same issue. So, it might be nice to have a call with ADCIRC developers soon. I still could not talk with Bob (from ESMF team) since he is busy with some other works but I will talk with him next week. So, we might wait for him and then talk with ADCIRC developers later. Anyway, let me know what you think?

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> , @Saeed Moghimi - NOAA Federal
***@***.***>
Ufuk thank you for the update. As long as this is the case, please have a
meeting with Bob as soon as possible to see
what can be done from the ESMF side. I will be at the ADCIRC User's meeting
next week (06/08 (T) - 06/09 (F)) and I will
try to talk to the developers over there. Could you please write a
paragraph or two (in this thread)  briefly explaining what is the problem
and what  we expect (want) the ADCIRC developers to help with. I understand
that everyone is busy, but somehow, we need to get them
more actively involved in helping, if ADCIRC is going to be one of the
model components in ufs-coastal.

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by saeed-moghimi-noaa** [comment]
@pvelissariou1 @uturuncoglu @danrosen25  

Thanks Takis and Ufuk. 

I totally agree. 

We are into the month six of the project and I feel like we need to move faster to make sure we cover the 1st year goals. Perhaps in parallel, NCAR team may start looking into ESMX driver implementation meanwhile waiting for developers to respond. Let me know what you think?

Thanks a lot,
-Saeed

**Comment by uturuncoglu** [comment]
@saeed-moghimi-noaa @danrosen25 is working on ESMX and the more stable version with new feature will be available in 8.5.0 (will be released in a month). Then, we could start to implement ESMX under ufs-coastal. Probably he has more information about it. In the meantime, I am plaining to keep working on ADCRIC cap to solve the issue. Since FVCOM and ADCIRC are very similar interns of mesh generation, I think FVCOM integration will be fast.

**Comment by dwirasae** [comment]
> @dwirasae As I see there are some elements with negative id (means ghost) but all nodes are resident. I am expecting to have positive element id for those elements. Do you have any explanation about it?

No, I do not. I expect those elements to have a postive id. Let me double check and get back to you.

**Comment by uturuncoglu** [comment]
@dwirasae Thanks for your help. Maybe there is a bug in the code. I'll double check in my side.

**Comment by uturuncoglu** [comment]
@dwirasae I am not sure I am following correctly or not but one example of having negative element id with all resident nodes from [florence_hsofs.atm2adc](https://github.com/noaa-ocs-modeling/CoastalApp-testsuite/tree/main/florence_hsofs.atm2adc) test. 

In the `PE0144/fort.18` file, I have following entry

```
FileFmt      1050624           0           0
NELG         3564104        7707        6972
       -6971
       -6972
       -6981
       -6982
       -6991
       -6992
       -7001
       -7002
       -7011
       -7012
       -7019
       -7020
       -7021
       -7022
       -7023
...
...
```

So, if you look at element with global id -7023 it is in 25th row (23th element locally since 2 rows are header). Then, if I look at `PE0144/fort.14` file to find its nodes, you just need to look at 23th entry and it is something like following,

`      23 3     4118     4119     4124`

So, local id of the element -7023 is 23 and seems that 4118, 4119 and 4124 are the nodes (I think these are global but all the entry in the element connection list positive). Again from the same file their entry like following,

```
    4118 -0.904350820000E+02  0.299931520000E+02  0.221426280000E+02 <--- 18th entry in NOMAD section
    4119 -0.904363980000E+02  0.299907540000E+02 -0.275970200000E+01 <--- 19th entry
    4124 -0.904338730000E+02  0.299905870000E+02 -0.298566510000E+01 <--- 24th entry
```

This does not say anything about the node is ghost or not since all ids are local and positive in `PE0144/fort.14`. Right? But, could find the global ids of those from `PE0144/fort.18` file by looking NNODG section. If I look at entry 18, 19 and 24 after NNODG section, you can find the global ids of those nodes and they are all positive. So, it seems that they are resident but if I look at RES NODE for other SEND/RECV PE section i could not find those nodes (maybe they are listed local ids). Anyway, let me know what you think?

**Comment by dwirasae** [comment]
@uturuncoglu. I look into the decomposition code. This could happen is a scenario where there are internal 'weir' boundaries (to handle levees, jetties).  The hsof grid may use these boundaries  in southern Louisiana.   I will take a look at the test and get back to you.

**Comment by uturuncoglu** [comment]
@dwirasae Thanks for your help. It is good to know, it is possible. Is there any way to identify those elements/nodes? BTW, following is the example for most elements (from PET144 and PET143). The pink region is exiting in the both PE and I am trying get rid of that. If I could identify those 'weir' boundaries, then maybe I could fix the issue. The subroutine to eliminate the ghost elements is ready but since I could not identify which element will be stay and which one will be eliminated, I could not make it work. Since this case uses 499 core, it is also hard to debug and follow the connection between elements and nodes. So, maybe it would be setup a realistic case with less number of cores and relatively low resolution mesh but that is another approach that we could take if I could not solve the issue. I will be out of office tomorrow but I'll check my mails and this issue page for any update.   

![Screenshot 2023-05-31 at 4 21 33 PM](https://github.com/oceanmodeling/ufs-coastal/assets/1228999/d8c9f248-8c6c-4593-90db-02118fcbbed8)

**Comment by uturuncoglu** [comment]
@dwirasae I look at it little bit more and I think that it is hard to access weir boundaries with the current version of code since this information is not written to any file like fort.14 or fort.18 and it is just available in adcprep tool. I am trying to include all the elements that at least two nodes are resident but in this case some nodes will be outside because those elements has only one resident nodes and ESMF mesh creation is complaining about it. At this point I am not sure those nodes related with weir boundaries or not until put some debug code to adcprep and try to write id of those nodes. Anyway, do you have any suggestion?

**Comment by uturuncoglu** [comment]
@dwirasae @saeed-moghimi-noaa @pvelissariou1 

As we discussed today we have two issues:

1. We need to eliminate ghost elements from the mesh. I tried different appraoches without any luck and just by looking element ids >0 is not working at this point. I think we need to have new data structures (simply mask) that indicates the element is ghost or not.

2. The import and export fields are need to be defined on ESMF_MESHLOC_ELEMENT (the default one is node). I think this is secondary issue and we need to solve the previous issue first. Then, we could look at this further.

Anyway, here is the details of my initial development to solve the (1) issue. 

I have just push the initial development in ADCIRC cap to resolve the issues to make it work with CMEPS mediator. It is not working at this point since we have no enough information to eliminate ghost elements and just looking to positive element ids is not solving. Here is the details of the code,

- I am calling a special routine called like `eliminate_ghosts()` just after reading mesh information. As you can see this depends on some configuration option called meshloc to isolate this development. The main goal is in here to remove the ghost elements from the every PEs and the mesh.

https://github.com/uturuncoglu/adcirc/blob/509cf9ad9551a4bfd7ca592a166e27ef5bd6420c/thirdparty/nuopc/adc_cap.F90#L686

- The routine is in the following link. The `ElMask` variable is just looking for the elements that `the_data%ElIDs(i) > 0` at this point and not working with CMEPS. So, we need to find a way in here to filter out the ghost elements.
 https://github.com/uturuncoglu/adcirc/blob/509cf9ad9551a4bfd7ca592a166e27ef5bd6420c/thirdparty/nuopc/adc_mod.F90#L424

**Comment by dwirasae** [comment]
@uturuncoglu Sorry for not responding promptly.  Do you have by any chance PEXXXX directories from running the prep?

**Comment by uturuncoglu** [comment]
@dwirasae No worries. I just switched another component - FVCOM to use my time more efficiently. Since that one shared most of the mesh generation code with ADCIRC, probably I'll have similar issue. Do you have access to NCAR's Cheyenne system. If so, you could find the files in the /glade/scratch/turuncu/COASTAL_APP/florence_hsofs.atm2adc_med directory. A this point I have only access Cheyenne and Orion.

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> Ufuk, is it possible to deploy all
these files on orion. ADCIRC folks have access to orion.

Thanks

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@dwirasae @pvelissariou1 I put the zipped PE files to `/work/noaa/nems/tufuk/PE.tar.gz` on Orion. Let me know if you have access issue.

**Comment by uturuncoglu** [comment]
@pvelissariou1 JFYI, I am getting following errors from ADCIRC, 

```
141: TIME STEP =  302402     ITERATIONS =    0     TIME =  0.60480400E+06
141:  ELMAX =  7.1438E+035 AT NODE     1267  SPEEDMAX =  3.3298E+034 AT NODE     1275  ON MYPROC =  125   ** WARNING: Elevation.gt.WarnElev **
141: TIME STEP =  302402     ITERATIONS =    0     TIME =  0.60480400E+06
141:  ELMAX =  7.1438E+035 AT NODE     1267  SPEEDMAX =  3.3298E+034 AT NODE     1275  ON MYPROC =  125
141:  ** ERROR: Elevation.gt.ErrorElev, ADCIRC stopping. **
156: TIME STEP =  302402     ITERATIONS =    0     TIME =  0.60480400E+06
156:  ELMAX =  4.2046E+276 AT NODE   199608  SPEEDMAX =     Infinity AT NODE   202232  ON MYPROC =  140   ** WARNING: Elevation.gt.WarnElev **
156: TIME STEP =  302402     ITERATIONS =    0     TIME =  0.60480400E+06
160: TIME STEP =  302402     ITERATIONS =    0     TIME =  0.60480400E+06
160:  ELMAX =  4.2046E+276 AT NODE  -199608  SPEEDMAX =     Infinity AT NODE  -202232  ON MYPROC =  144   ** WARNING: Elevation.gt.WarnElev **
160: TIME STEP =  302402     ITERATIONS =    0     TIME =  0.60480400E+06
160:  ELMAX =  4.2046E+276 AT NODE  -199608  SPEEDMAX =     Infinity AT NODE  -202232  ON MYPROC =  144
160:  ** ERROR: Elevation.gt.ErrorElev, ADCIRC stopping. **
156:  ELMAX =  4.2046E+276 AT NODE   199608  SPEEDMAX =     Infinity AT NODE   202232  ON MYPROC =  140
156:  ** ERROR: Elevation.gt.ErrorElev, ADCIRC stopping. **
```

This is without CMEPS and I already updated my fork with ADCIRC/development. Since we are using different dataset to force ADCIRC under ufs-coastal, it would create issue. anyway, I think you are working on CoastalApp testsuite to make input files consistent and adjust configuration to write output regular intervals. So, once this is done, I think I need to update all the RTs defined under ufs-coastal and again compare them with the CoastalApp testsuite ones.

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> Ufuk, I was getting these
warnings/errors when running PaHM+ADCIRC due to a bug in PaHM. What test
case are you running?

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 This is just ADCIRC coupled with CDEPS.

**Comment by pvelissariou1** [comment]
ok, thanks. I'll check it out.

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 I might need input for CDEPS to use under `ike_shinnecock.atm2adc` case. I checked `ike_shinnecock.atm2sch` but that one is using different start date for the simulation and it does not cover this configuration. Anyway, I'll try to force it with ERA5 and let you know.

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> Please Let me know, I can create an
atm. input file if needed.

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 Okay. Downloading now. It would be very coarse in terms of spatial resolution but at least it will give me chance to run the case. We could update the input later once we have consistent inputs for all Ike and Florence. I'll let you know if I need help about input.

**Comment by uturuncoglu** [comment]
@pvelissariou1 I am getting following error again from this case,

```
2: TIME STEP =  540535     ITERATIONS =   20     TIME =  0.10810700E+07
2:  ELMAX =  1.3245E+003 AT NODE      -98  SPEEDMAX =  1.2296E+003 AT NODE      -25  ON MYPROC =    2
9: TIME STEP =  540535     ITERATIONS =   20     TIME =  0.10810700E+07
9:  ELMAX =  1.3245E+003 AT NODE       98  SPEEDMAX =  1.6949E+003 AT NODE       26  ON MYPROC =    9   ** WARNING: Elevation.gt.WarnElev **
9: TIME STEP =  540535     ITERATIONS =   20     TIME =  0.10810700E+07
9:  ELMAX =  1.3245E+003 AT NODE       98  SPEEDMAX =  1.6949E+003 AT NODE       26  ON MYPROC =    9
9:  ** ERROR: Elevation.gt.ErrorElev, ADCIRC stopping. **
2:  ** ERROR: Elevation.gt.ErrorElev, ADCIRC stopping. **
```

Maybe it is because of the input. Not sure. Any idea?

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> It is strange. Let me check this out.
Please share the atmospheric data file.
Need to go to get my car from the shop. I'll test this later tonight. I'll
keep you posted.

Takis

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 It is in following directory on Cheyenne. 

`/glade/scratch/turuncu/COASTAL_APP/ike_shinnecock.atm2adc/INPUT/download_200809_rlat.nc`

Maybe I could try to write out field in ADCIRC side in VTK format using ESMF calls to check them.

**Comment by uturuncoglu** [comment]
@pvelissariou1 I think the interpolation is not working properly. The ADCRIC that I used in here is using connectors and no modification in the cap. The following image is the decomposition blocks,

![Screenshot 2023-07-07 at 3 30 28 PM](https://github.com/oceanmodeling/ufs-coastal/assets/1228999/74db3ba2-5108-479e-9618-c840aaf71048)

and the next one is pressure. As you can see that we have issue in the interpolation in the boundaries between decomposition elements. This could be caused by ghost elements. I just wonder that did you check the results under CoastalApp-test suite? 

![Screenshot 2023-07-07 at 3 32 06 PM](https://github.com/oceanmodeling/ufs-coastal/assets/1228999/74753860-c23e-4659-b2f1-0c08825d5260)

In the CoastalApp-testsuite, the run sequence is defined like following,

```
# Run Sequence #
runSeq::
  @3600
    ATM -> OCN   :remapMethod=redist
    ATM
    OCN
  @
::
```

but in my case, I am using following to force bit-to-bit reproducibility and in my case I am not using data that is in the same resolution with the model and using bilinear interpolation (in the CoastalApp case it is redist),

```
runSeq::
  @3600
    ATM -> OCN :remapMethod=bilinear:unmappedaction=ignore:zeroregion=select:srcTermProcessing=0:termOrder:srcseq
    ATM
    OCN
  @
::
```

Anyway, I'll do further investigation and let you know.

**Comment by uturuncoglu** [comment]
@pvelissariou1 I am trying to run CoastalApp-testsuite on Orion to see I could reproduce the issue in there. I'll update you.

**Comment by uturuncoglu** [comment]
@pvelissariou1 Okay. I am getting following error. I think there is an issue in the system or related with the nems account that I am using since I think there was a disk quota issue recently. Is it possible to add me to the coastal group on Orion? I could also try on Cheyenne again but I am not sure CoastalApp-testsuite is working in there or not.

```
[1688769215.647710] [Orion-01-09:440986:0]        ib_verbs.h:88   UCX  ERROR ibv_exp_query_device(mlx5_0) returned 38: No space left on device
[1688769215.647753] [Orion-01-09:440984:0]        ib_verbs.h:88   UCX  ERROR ibv_exp_query_device(mlx5_0) returned 38: No space left on device
```

**Comment by uturuncoglu** [comment]
@pvelissariou1 Just for checking, I run SCHISM for the same domain (`ike_shinnecock.atm2sch` configuration) and forced with the CDEPS (again ERA5) but I am not seeing same issue. So, it seems that ghost elements are creating issue in the interpolation (at this point I am not sure why we are seeing issue also inside of the elements but those could be side effects). I still could not run same case (`ike_shinnecock.atm2adc`) under CoastalApp and still getting same error. I'll try again on Cheyenne to see what happens. If the all cases with ADCIRC uses same grid for all the components maybe this is not matched before but not sure.

**Comment by uturuncoglu** [comment]
@pvelissariou1 @saeed-moghimi-noaa  I could able to run out-of-box `ike_shinnecock.atm2adc` ADCIRC case on Orion from CoastalApp-test suite and it seems that it has also same issue. See following plot for surface pressure. Also note that this uses redist since it uses ATMMESH. I am not sure why this error is not cached before under CoastalApp or their applications. Any idea? Definitely ADCIRC cap has issue due to ghost elements. This is the connector only and I think I could add it to the presentation as issue 3. Let me know what you think? If you could reproduce the same in your end that would be great.

![Screenshot 2023-07-07 at 11 30 07 PM](https://github.com/oceanmodeling/ufs-coastal/assets/1228999/05f227c3-9018-416e-864e-1243f3d0cf3a)

**Comment by uturuncoglu** [comment]
@pvelissariou1 Based on my experience it is not likely but this could be an initialization issue of the configuration too. It worths to get opinion from ADCIRC developers. But, still strange to see this pattern in the input forcing like pressure field after interpolation. The mode does not involve that step. So, this seems related with the interpolation artifact due to the ghost element. I'll check it with Bob from ESMF team and get his feeling too. In any case, it would be great that someone else could confirm this finding in CoastalApp side.

**Comment by uturuncoglu** [comment]
@pvelissariou1 I think the issue that I am seeing with original code could be related to the halo update. Since I am outputting fields just before/after run call with esmf field write, I might miss that halo update call that is inside of the adcirc run. So this might explain the structure seen in the connector only (and using redist) run under CoastalApp test suite. I'll check the model output to be sure. I am also planing to test same case by changing redist to bilinear to see any change in the result.

**Comment by pvelissariou1** [comment]
@Ufuk Turuncoglu ***@***.***> Thank you for the updates. I ran the
test case on Orion and I didn't see the issues you described in the output
file (fort.73.nc) or the input file. Also, in CoastalApp I haven't updated
ADCIRC to its very latest version.

*NOTE:* FYI, on Orion using *intel/2019.5* sometimes the simulation goes
through and sometimes produces MPI errors as you mentioned before.
*intel/2020.2* does not work due to issues with stdc++ library for the CXX
api. *intel/2022.1.2* works flawlessly (I am switching CostalApp to use
this compiler on Orion). Different HPC clusters have different
configurations and things might break due to misconfigurations, for example
intel (19/2019) works very well on
*hera* and *Frontera* but it doesn't work as expected on *Orion* and
*ParallelWorks*.

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 Okay. At this point, I am assuming CoastalApp is working fine even without using latest version of ADCIRC. I also tried to run `ike_shinnecock.atm2adc` case under CoastalApp but changing redist option to bilinear to see any issue in the interpolation but it fails with `11: NetCDF Error: Variable col in weightmatrix_ATM-TO-OCN_inst_merid_wind_height10m.nc : NetCDF: Name contains illegal characters`. So, maybe ATMMESH is not designed to work like that. Anyway, the issue that I am seeing under UFS-Coastal is maybe a specific issue. It is hard to create same configuration under CoastalApp since we replaced ATMMESH with CDEPS. I'll look at more to find the source of the interpolation error. BTW, even if you have correct fort.73.nc file, it does not mean the coupling is working fine since the import and export states are showing zero along the ghost regions and those are transferred to other components. We are using ATMMESH and there is nob two way coupling in here but in two-way coupled application, it would be nice to check the receiver about those zeros along the boundary to be sure those are not transferred.

**Comment by uturuncoglu** [comment]
@pvelissariou1 Thanks for your help. I think it would be great to test ufs-coastal with the same input used by `ike_shinnecock.atm2adc` configuration under CoastalApp testsuite. Is there any progress to make input files consistent across the CoastalApp testsuite and having those files in their original grid for CDEPS?

**Comment by uturuncoglu** [comment]
@pvelissariou1 I added new test to RTs. It is `coastal_ike_shinnecock.atm2adc` and working with CMEPS. Please try it under Orion and let me know if you have any issue. Same RT folder can be used.

**Comment by pvelissariou1** [comment]
@uturuncoglu Thank you, I will try it. Tomorrow I'll update the nuopc cap for FVCOM as well.

**Comment by uturuncoglu** [comment]
@pvelissariou1 @saeed-moghimi-noaa We were discussing with @dwirasae last couple of days related to the issue in the ADCIRC. We might have a call with him today. Let me know if you want to join.

**Comment by pvelissariou1** [comment]
Ufuk, thank you, I will join any time today. FYI, Saeed is extremely busy
and I don't know if he will have free time to join.

Panagiotis Velissariou, Ph.D., P.E.
UCAR Scientist
National Ocean and Atmospheric Administration
National Ocean Service
Office of Coast Survey CSDL/CMMB
Physical Scientist - Project Lead
cell: (205) 227-9141
email: ***@***.***

**Comment by uturuncoglu** [comment]
@pvelissariou1 I am trying to update `florence_hsofs.atm2adc` case from Coastal App test suite but some files are missing in the following directory such as fort.14. Is this case running under test suite? I could keep RTs under ufs-cosatal same and not update. Let me know your preference.

https://github.com/noaa-ocs-modeling/CoastalApp-testsuite/tree/develop/florence_hsofs.atm2adc/run

**Comment by pvelissariou1** [comment]
@uturuncoglu Ufuk, those "missing files" are contained in a separate archive accessible via the command:
``wget https://tacc-nos-coastalapp-testsuit.s3.amazonaws.com/hsofs-data-v3.tgz``
Over there you will find the files for adcirc. The files are rather large, that is why I don't have them in the github repo.
You might want to extract the files from the archive and put them in the RT case in UFS-Coastal.

**Comment by uturuncoglu** [comment]
@pvelissariou1 Okay. Let me check it. Thanks.

**Comment by uturuncoglu** [comment]
@pvelissariou1 I think the file is under CoastalApp-testsuite/comm/mesh/. Right? Since there is no any subfolder dedicated to the component, I assuming that these files belongs to ADCIRC.

**Comment by pvelissariou1** [comment]
Correct

**Comment by uturuncoglu** [comment]
@pvelissariou1 Do we have 2d version of `Florence_HWRF_HRRR_HSOFS.nc` that can be used under CDEPS? Is it possible to create it?

**Comment by pvelissariou1** [comment]
@uturuncoglu I think I had created that file before. Let me check and if not I will create it.

**Comment by uturuncoglu** [comment]
@dwirasae @pvelissariou1 @saeed-moghimi-noaa There is a bug in ADCIRC side that prevents me to use different set of PEs. If I set it the nodes coordinates goes wrong and creates issue. If I align processors with other components it works as expected. Anyway, I am trying to debug and fix at this point. Another aspect is that ADCIC has capability to use dedicated I/O tasks but this feature can nat be used with the existing implementation of the NUOPC cap. We have same issue with SCHSIM new I/O feature. Anyway, I think they are not high priority at this point but it would be nice to open a ticket in component repository side to track.

**Comment by uturuncoglu** [comment]
@dwirasae @pvelissariou1 @saeed-moghimi-noaa In addition to the previous issue which I am still trying to fix. I think ADCIRC cap also needs to be fixed like FVCOM. I check the ADCIRC output with adcircpy tool and did not see any issue in the ADCIRC output but I believe that the way to get the data from input states and fill ADCIRC internal data structures is wrong. It still assumes that the data coming from ESMF mesh is on nodes. In fact, the ESMf fields on elements at this point.

**Comment by uturuncoglu** [comment]
@pvelissariou1 I am testing to use 1d data (on model mesh) through the CMEPS. `data/atm/Florence_HWRF_HRRR_HSOFS.nc` is used in atm2adc Florence case. This file covers from 2018-09-09 06:00:00 to 2018-09-30 23:00:00 but if you check following file https://github.com/noaa-ocs-modeling/CoastalApp-testsuite/blob/develop/florence_hsofs.atm2adc/run/model_configure, it seems that simulation is starting from 2018-09-07 00:00:00. So, CDEPS is failing since it could not find the required data in the file. If you don't mind could you double check. We might need to extend the data to cover simulation period. Maybe I need to download the data from scratch if you already updated. Anyway, let me know what you think.

**Comment by uturuncoglu** [comment]
@pvelissariou1 I am not sure but there might be an issue with `Florence_HWRF_HRRR_HSOFS.nc` input file. Here is the snapshot from ncdump command. At this point, number of element is greater than number of nodes. How this could happen. Is this file also includes ghost elements? Even if that I think it is way more than the number of nodes. I am not sure what is wrong with it.

```
netcdf Florence_HWRF_HRRR_HSOFS {
dimensions:
	level = 1 ;
	node = 1813443 ;
	element = 3564104 ;
	time = 522 ;
	noel = 3 ;
```

I have also case of unstructured data used by CESM and in their case the mesh file has following,

```
netcdf ne30pg3_ESMFmesh_cdf5_c20211018 {
dimensions:
	nodeCount = 48602 ;
	elementCount = 48600 ;
	maxNodePElement = 4 ;
	coordDim = 2 ;
```

So, as I expected the number of element is less than number of nodes. Anyway, maybe ATMMESH has some feature and required data like this but I think it would be hard to use it under CDEPS since there is inconsistency between node and element counts. We have case under CESM that uses unstructured data and it is working without any issue. I am modifying data file in a same way but in our case CDEPS fails. Anyway, it would be nice to check the data file and see why elements are greater than the nodes.

**Comment by uturuncoglu** [comment]
@pvelissariou1 I updated ADCIRC under ufs-coastal. Now, it points our official fork: https://github.com/oceanmodeling/adcirc and we are using `feature/coastal_app` branch for development.

**Comment by uturuncoglu** [comment]
@pvelissariou1 Do you have any update about `Florence_HWRF_HRRR_HSOFS.nc`file? Before testing under CDEPS, I think we have to figure our why this file has way more element than the nodes.

**Comment by pvelissariou1** [comment]
@uturuncoglu The number of elements and the number of nodes are correct. They much to other fort.14 HSOFS data files. What is the issue here?

**Comment by uturuncoglu** [comment]
@pvelissariou1 I am not sure but maybe number of elements also includes ghost elements (since you are creating ti from fort.14) and that creates issue when we are trying to create ESMF Mesh from the data file to sue it under CDEPS. Is it possible to remove the ghost elements from the data file. Otherwise, we need to find a way.

**Comment by pvelissariou1** [comment]
@uturuncoglu , @yunfangsun , @saeed-moghimi-noaa UFS-Coastal fails to compile using the latest versions of ADCIRC and WW3 due to name conflict of the "MODULE CONSTANTS" defined in both ADCIRC and WW3.
Workaround: rename "MODULE CONSTANTS" to "MODULE ADC_CONSTANTS"  in src/constants.F90 and all occurences of "USE CONSTANTS" to "USE ADC_CONSTANTS". I will submit a PR to adcirc repo.

**Comment by pvelissariou1** [comment]
@uturuncoglu , @yunfangsun , @saeed-moghimi-noaa Need to freeze on latest working vesion/commit of all models and go through RTs when a specific model component is to be updated.

**Comment by pvelissariou1** [comment]
@uturuncoglu , @yunfangsun , @saeed-moghimi-noaa Zach has merged my PR to the main. This PR solved the issue of name conflicts with WW3. @uturuncoglu Ufuk please update the adcirc fork in oceanmodeling to reflect these changes. Note that there is no development branch in ADCIRC anymore.

**Comment by pvelissariou1** [comment]
@uturuncoglu ADCIRC fails to compile when the flag -DBUILD_UTILITIES=ON is used. It seems that the Fortran_LINELENGTH_FLAG setting (-132) does not propagate to the utility folders (e.g., owi22). Compiling ADCIRC directly from ADCIRC-interface/ADCIRC using cmake, produces all executables properly. While this is not urgent, we need to fix it from the top level cmake config files.

**Comment by uturuncoglu** [comment]
@pvelissariou1 Let me sync and fix the issues. I am in ESMf core team call this morning but I could make it in the afternoon.

**Comment by pvelissariou1** [comment]
@uturuncoglu No rush. I am using for my simulations the modified ufs-coastal at this point.

**Comment by pvelissariou1** [comment]
@uturuncoglu I have successfuly compiled the latest version of ADCIRC in UFS-Coastal. Minor changes were required in ADCIRC-interface CMake files and in ADCIRC/thirdparty/nuopc. I'll create a PR for the UFS-Coastal changes and separate RPs for the ADCIRC modifications. We need to push Dam's modifications on adcprep upstream. @dwirasae Dam, I think so far your adcprep modifications work, is it ok if create a PR to ADCIRC for those or you prefer to do it?

**Comment by uturuncoglu** [comment]
@pvelissariou1 Okay. That is great. Please create PR. I have meeting until noon. I'll check it later today.

**Comment by pvelissariou1** [comment]
@uturuncoglu I'll make a PR to UFS-Coastal and another one to ADCIRC upstream. Let's have a quick chat after your meeting.

**Comment by dwirasae** [comment]
Please feel free to go ahead and create PR on adcprep.

**Comment by pvelissariou1** [comment]
@dwirasae Already working on it, re-testing. I'll submit the PR shortly. Thank you so much.
